keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,886 @@
+    - `pod.spec.hostIP`
+    - `pod.spec.hostIPs[]`
+    - `pod.spec.podIP`
+    - `pod.spec.podIPs[]`
+    - `service.spec.clusterIP`
+    - `service.spec.clusterIPs[]`
+    - `service.spec.externalIPs[]`
+    - `service.spec.loadBalancerSourceRanges[]`
+    - `service.status.loadBalancer.ingress[].ip`
+
+  - in `networking`:
+    - `ingress.status.loadBalancer.ingress[].ip`
+    - `networkpolicy.spec.egress[].to[].ipBlock.cidr`
+    - `networkpolicy.spec.egress[].to[].ipBlock.except[]`
+    - `networkpolicy.spec.ingress[].from[].ipBlock.cidr`
+    - `networkpolicy.spec.ingress[].from[].ipBlock.except[]`
+    - `serviceCIDR.spec.cidrs[]`
+
+  - in `discovery`:
+    - `endpointslice.endpoints[].addresses[]`
+
+#### Updates to pre-existing objects with pre-existing invalid fields
+
+We want to allow making changes to objects containing invalid
+IPs/CIDRs, as long as the IPs/CIDRs themselves are not changed.
+However, in the case of array-valued fields, it seems best to allow
+adding and removing new valid IPs/CIDRs without needing to
+simultaneously fix pre-existing invalid IPs/CIDRs.
+
+So, the rule will be rougly: "When validating an Update, any IP/CIDR
+that exists in the old version of the field is allowed to exist in the
+new version of the field." For NetworkPolicy, that is extended to "any
+CIDR that exists anywhere in the old version of the object is allowed
+to exist anywhere in the new version of the object" (to allow
+inserting new rules into a policy without having to fix invalid CIDRs
+in unrelated later rules).
+
+However, `Endpoints` and `EndpointSlice` will be treated differently,
+because (a) they are large enough that doing additional validation on
+them could be noticeably slow, (b) in most cases, they are generated
+by controllers that only write out valid IPs anyway, (c) in the case
+of `EndpointSlice`, if we were going to allow moving bad IPs around
+within a slice without revalidation, then we ought to allow moving
+them between related slices too, which we can't really do.
+
+So for `Endpoints` and `EndpointSlice`, the rule will be that invalid
+IPs are only allowed to remain unfixed if the update leaves the entire
+`.subsets` / `.addresses` unchanged. So you can edit the labels or
+annotations without fixing invalid endpoint IPs, but you can't add a
+new IP while leaving existing invalid IPs in place.
+
+#### Plan for allowing updates to immutable fields
+
+Four of the fields listed above are immutable:
+
+  - `pod.spec.dnsConfig.nameservers[]`
+  - `pod.spec.hostAliases[].ip`
+  - `service.spec.clusterIP`
+  - `service.spec.clusterIPs[]`
+
+For these fields we will add the special rule that you are allowed to
+modify them if:
+
+  - the old value does not pass current validation rules, _and_
+  - the new value is the canonical representation of the old value.
+
+So given `clusterIP: 172.030.099.099`, you would be allowed to modify
+it to `clusterIP: 172.30.99.99`, but not to any other value. (For
```

Comment:
Yes, we _could_ do that; @thockin proposed that we allow fixing immutable values [here](https://github.com/kubernetes/kubernetes/pull/107552#issuecomment-1035630181) but that was the first reference to immutable values in that thread, so I'm not sure if there was discussion somewhere else that led to that, or if this was inspired by previous discussions of validation ratcheting in other contexts or what...

Reply:
no other discussion I'm aware of, but assuming the time it has passed since golang 1.17 and most of the kubernetes ecosystem uses golang, I have doubts there are any of these weird apis out there at this time, and it does not seem worth the effort vs having to add more complex logic on the Service machinery ... specially around about ClusterIP and the immutable properties